From 85feac946298302b5d7e202e6a7fc0dc58b76f98 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Thu, 28 Aug 2014 03:06:38 -0400
Subject: [pnp] pnpacpi: Fix acpi_pnp_match()

Message-id: <20140828030638.2117.1286.stgit@gir.stowe>
Patchwork-id: 89318
O-Subject: [RHEL7.1 PATCH 30/30] ACPI / PNP: Fix acpi_pnp_match()
Bugzilla: 1128632
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1128632
Upstream Status: b6328a07bd6b3d31b64f85864fe74f3b08c010ca

commit b6328a07bd6b3d31b64f85864fe74f3b08c010ca
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Wed Jul 30 00:23:09 2014 +0200

    ACPI / PNP: Fix acpi_pnp_match()

    The acpi_pnp_match() function is used for finding the ACPI device
    object that should be associated with the given PNP device.
    Unfortunately, the check used by that function is not strict enough
    and may cause success to be returned for a wrong ACPI device object.

    To fix that, use the observation that the pointer to the ACPI
    device object in question is already stored in the data field
    in struct pnp_dev, so acpi_pnp_match() can simply use that
    field to do its job.

    This problem was uncovered in 3.14 by commit 202317a573b2 (ACPI / scan:
    Add acpi_device objects for all device nodes in the namespace).

    Fixes: 202317a573b2 (ACPI / scan: Add acpi_device objects for all device nodes in the namespace)
    Reported-and-tested-by: Vinson Lee <vlee@twopensource.com>
    Cc: 3.14+ <stable@vger.kernel.org> # 3.14+
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pnp/pnpacpi/core.c b/drivers/pnp/pnpacpi/core.c
index eb415b2..b9c7f48 100644
--- a/drivers/pnp/pnpacpi/core.c
+++ b/drivers/pnp/pnpacpi/core.c
@@ -338,8 +338,7 @@ static int __init acpi_pnp_match(struct device *dev, void *_pnp)
  struct pnp_dev *pnp = _pnp;
 
  /* true means it matched */
- return !acpi->physical_node_count
-     && compare_pnp_id(pnp->id, acpi_device_hid(acpi));
+ return pnp->data == acpi;
 }
 
 static int __init acpi_pnp_find_device(struct device *dev, acpi_handle * handle)
-- 
1.7.1